sonar-batch-protocol/src/main/java/org/sonar/batch/protocol/ProtobufUtil.java

```diff
@@ -19,72 +19,103 @@
     } finally {
       IOUtils.closeQuietly(input);
     }
   }
 
+  /**
+   * Writes a single message to a file, by replacing existing content. The message is
+   * NOT appended.
+   */
   public static void writeToFile(Message message, File toFile) {
     OutputStream out = null;
     try {
       out = new BufferedOutputStream(new FileOutputStream(toFile, false));
       message.writeTo(out);
     } catch (IOException e) {
-      throw new IllegalStateException("Unable to write protocol buffer data to file " + toFile, e);
+      throw new IllegalStateException(format("Unable to write protobuf message to file %s", toFile), e);
     } finally {
       IOUtils.closeQuietly(out);
     }
   }
 
-  public static void appendToFile(Message message, File toFile) {
+  public static void writeStreamToFile(Iterable<? extends Message> messages, File toFile, boolean append) {
     OutputStream out = null;
     try {
-      out = new BufferedOutputStream(new FileOutputStream(toFile, true));
-      message.writeDelimitedTo(out);
+      out = new BufferedOutputStream(new FileOutputStream(toFile, append));
+      for (Message message : messages) {
+        message.writeDelimitedTo(out);
+      }
     } catch (IOException e) {
-      throw new IllegalStateException("Unable to append protocol buffer data to file " + toFile, e);
+      throw new IllegalStateException(format("Unable to write protobuf messages to file %s", toFile), e);
     } finally {
       IOUtils.closeQuietly(out);
     }
   }
 
-  public static <MESSAGE extends Message> void writeMessagesToFile(Iterable<MESSAGE> messages, File file) {
-    OutputStream out = null;
+  public static <MSG extends Message> CloseableIterator<MSG> readStreamFromFile(File file, Parser<MSG> parser) {
     try {
-      out = new BufferedOutputStream(new FileOutputStream(file, true));
-      for (MESSAGE message : messages) {
-        message.writeDelimitedTo(out);
-      }
-    } catch (IOException e) {
-      throw new IllegalStateException("Failed to read file: " + file, e);
-    } finally {
-      IOUtils.closeQuietly(out);
+      return new ProtobufIterator<>(parser, new BufferedInputStream(new FileInputStream(file)));
+    } catch (FileNotFoundException e) {
+      throw new IllegalStateException(format("Unable to read protobuf file %s", file), e);
     }
   }
 
+  private static class ProtobufIterator<MSG extends Message> extends CloseableIterator<MSG> {
+    private final Parser<MSG> parser;
+    private final InputStream input;
+
+    private ProtobufIterator(Parser<MSG> parser, InputStream input) {
+      this.parser = parser;
+      this.input = input;
+    }
+
+    @Override
+    protected MSG doNext() {
+      try {
+        return parser.parseDelimitedFrom(input);
+      } catch (InvalidProtocolBufferException e) {
+        throw new IllegalStateException(e);
+      }
+    }
+
+    @Override
+    protected void doClose() throws Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Protected method 'doClose(...)' is never used. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUnusedProtectedMethod)